include/deal.II/matrix_free/vector_data_exchange.h

```diff
@@ -247,6 +247,307 @@ namespace internal
+          const ArrayView<const Number> &             locally_owned_array,
+          const std::vector<ArrayView<const Number>> &shared_arrays,
+          const ArrayView<Number> &                   ghost_array,
+          const ArrayView<Number> &                   temporary_storage,
+          std::vector<MPI_Request> &                  requests) const;
+
+        template <typename Number>
+        void
+        export_to_ghosted_array_finish_impl(
+          const ArrayView<const Number> &             locally_owned_array,
+          const std::vector<ArrayView<const Number>> &shared_arrays,
+          const ArrayView<Number> &                   ghost_array,
+          std::vector<MPI_Request> &                  requests) const;
+
+        template <typename Number>
+        void
+        import_from_ghosted_array_start_impl(
+          const VectorOperation::values               vector_operation,
+          const unsigned int                          communication_channel,
+          const ArrayView<const Number> &             locally_owned_array,
+          const std::vector<ArrayView<const Number>> &shared_arrays,
+          const ArrayView<Number> &                   ghost_array,
+          const ArrayView<Number> &                   temporary_storage,
+          std::vector<MPI_Request> &                  requests) const;
+
+        template <typename Number>
+        void
+        import_from_ghosted_array_finish_impl(
+          const VectorOperation::values               vector_operation,
+          const ArrayView<Number> &                   locally_owned_storage,
+          const std::vector<ArrayView<const Number>> &shared_arrays,
+          const ArrayView<Number> &                   ghost_array,
+          const ArrayView<const Number> &             temporary_storage,
+          std::vector<MPI_Request> &                  requests) const;
+
+        template <typename Number>
+        void
+        reset_ghost_values_impl(const ArrayView<Number> &ghost_array) const;
+
+      private:
+        /**
+         * Global communicator.
+         */
+        MPI_Comm comm;
+
+        /**
+         * Shared-memory sub-communicator.
+         */
+        MPI_Comm comm_sm;
+
+        /**
+         * Number of locally-owned vector entries.
+         */
+        unsigned int n_local_elements;
+
+        /**
+         * Number of global vector entries.
+         */
+        types::global_dof_index n_global_elements;
+
+        /**
+         * Number of ghost vector entries.
+         */
+        unsigned int n_ghost_elements;
+
+        /**
+         * Remote ranks from where data is received.
+         */
+        std::vector<unsigned int> recv_remote_ranks;
+
+        /**
+         * Pointer into the receive buffer.
+         */
+        std::vector<std::pair<types::global_dof_index, types::global_dof_index>>
+          recv_remote_ptr;
+
+        /**
+         * TODO
```

Comment:
Could you please fill in this variable, along with the other todos further down in this file?